pipelines/cogito-acceptance.yml

```diff
@@ -0,0 +1,68 @@
+# NOTE
+# This pipeline doesn't have the standard handlers (on_success, on_failure, on_error,
+# on_abort) because it is meant to be run by tests for cogito itself.
+# The handler output just adds noise to the tests.
+
+###########################################################
+
+resource_types:
+
+  - name: cogito
+    type: registry-image
+    check_every: 24h
+    source:
+      repository: pix4d/cogito
+      tag: ((cogito-tag))
+
+###########################################################
+
+resources:
+
+  - name: cogito
+    type: cogito
+    check_every: never
+    source:
+      log_level: debug
+      owner: ((github-owner))
+      repo: ((repo-name))
+      access_token: ((oauth-personal-access-token))
+      gchat_webhook: ((gchat_webhook))
+      notify_on_states:  [abort, error, failure, pending, success]
+
+  - name: target-repo.git
+    type: git
+    source:
+      uri: https://github.com/((github-owner))/((repo-name)).git
+      branch: ((target-branch))
+
+  - name: cogito-repo.git
+    type: git
+    source:
+      uri: https://github.com/Pix4D/cogito.git
+      branch: ((cogito-branch))
+
+###########################################################
+
+jobs:
+
+  - name: chat-message-append.job
+    plan:
+      - get: target-repo.git
+      - put: cogito
+        inputs: [target-repo.git]
+        params:
+          state: success
```

Comment:
So this also pushes the status to github?

Just for my understanding: let say that this put step fails, but after it managed to push success status to github i.e while posting msg to chat. What would happen? In other pipelines we would have on_failure step, so it would post again the status and default build msg, hence overwriting the status from this put step?

Reply:
Good question.

First, I want to make it clear that I don't have strong opinions on the current behavior. We can change it if we think there is a better way.

Before answering your question, some context. The "problem" is two-fold, and I think that I mentioned it in the design document for this ticket:

1. A `put` on a Concourse resource doesn't know _where_ is running: it could be running at the end of a job via an handler or it could be running in the middle of a job as an explicit `put` after a task. It doesn't even know the build state (see https://github.com/concourse/concourse/issues/8541).
2. Currently we require to always specify the `state`, also when arguably it doesn't make a lot of sense (like in this case).

To answer the question, the sequence should be the following:

a. current job marked as success in github
b. posting message to chat fails, cogito exits 1, concourse considers the job errored (color: orange)
c. if there is an handler, cogito runs again and (if everything goes well) sets the same job as errored
d. if there is no handler, the job remains forever as success on github

Note that something similar can already happen:

1. job starts, cogito sets github state as pending
2. a task exits 1
3. concourse runs the on_failure hook
4. cogito runs and for any reasons it fails to update the state
5. the job remains forever in state pending on github

This makes me realize that we should at the very minimum explain in the cogito readme (we cannot enforce) that a `put` in the middle of a job should have only state `pending`, because in case of subsequent malfunctioning it will be "less wrong" than a `success`, so thanks!

Another somewhat related change I am considering is to do like the hangouts resource, that it always exits 0 also on error, and sets the metadata fields:
```
status	Failed                                <== as resource output metadata
error	Missing 'webhook_url' in source       <== as resource output metadata
```
which might be better?

Reply:
> posting message to chat fails, cogito exits 1, concourse considers the job errored (color: orange)

why errored? Errored is if volume disappears or some other kind of Concourse related issue. This will be red and marked as failure?
Or am I wrong?

Reply:
Each time I am confused, so I look at the table in the Cogito README and I _assume_ that it is correct. I created it based on Concourse documentation:

- failure | 🔴 - task exited non 0
- error | 🟠 - any other error besides failure or abort (pipeline configuration error, network error, timeout, ...)

it is true that the resource itself if it exits non 0 it is reported red/failure, but I _assume_ that the job becomes orange/error. I might be wrong.